fed_avg/server.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import annotations
+
+import random
+from typing import Any, Callable, Iterable, Sequence
+
+import torch
+
+from cezo_fl.shared import CriterionType
+from cezo_fl.util.metrics import Metric
+from cezo_fl.util.model_helpers import model_forward
+
+from .client import FedAvgClient
+
+
+class FedAvgServer:
+    def __init__(
+        self,
+        clients: Sequence[FedAvgClient],
+        device: torch.device,
+        server_model: torch.nn.Module,
+        server_criterion: CriterionType,
+        server_accuracy_func: Callable,
+        num_sample_clients: int = 10,
+        local_update_steps: int = 10,
+    ) -> None:
+        self.clients = clients
+        self.device = device
+        self.num_sample_clients = num_sample_clients
+        self.local_update_steps = local_update_steps
+
+        self.server_model = server_model
+        self.server_criterion = server_criterion
+        self.server_accuracy_func = server_accuracy_func
+
+        self.dtype = next(server_model.parameters()).dtype
+
+    def get_sampled_client_index(self) -> list[int]:
+        return random.sample(range(len(self.clients)), self.num_sample_clients)
+
+    def set_learning_rate(self, lr: float) -> None:
+        # Client
+        for client in self.clients:
+            for p in client.optimizer.param_groups:
+                p["lr"] = lr
+        # Server
+        if self.server_model:
+            for p in self.optim.param_groups:
+                p["lr"] = lr
+
+    def update_server_model(self, running_sum: Sequence[torch.Tensor]) -> None:
+        self.server_model.train()
+        with torch.no_grad():
+            for p, to_set_p in zip(self.server_model.parameters(), running_sum):
+                p.set_(to_set_p.div_(self.num_sample_clients))
+
+    def train_one_step(self) -> tuple[float, float]:
+        # Step 0: initiate something
+        sampled_client_indices: list[int] = self.get_sampled_client_index()
+
+        # Step 1 & 2: pull model and local update
+        running_sum = [0 for _ in self.server_model.parameters()]
+        step_train_loss = Metric("train_loss")
+        step_train_accuracy = Metric("train_loss")
+        for index in sampled_client_indices:
+            client = self.clients[index]
+            client.pull_model(self.server_model)
+            client_loss, client_accuracy = client.local_update(self.local_update_steps)
+            step_train_loss.update(client_loss)
+            step_train_accuracy.update(client_accuracy)
+            for i, p in enumerate(client.model.parameters()):
+                running_sum[i] += p.to(self.device)
+
+        self.update_server_model(running_sum)
```

Comment:
I would suggest to make this out. Make the function like this
```python
self.aggregate_client_models([self.client[index] for index in sampled_client_indices])
```
Benefits are
1. This will make the future RPC or parallelism refactor cleaner 
2. Move the `running_sum` into sub-rountine so that the existence time is shorter
3. Cleaner structure and more flexible for future other federated learning extension.

Reply:
@BichengYing 
Let me know if we want to do the following:

the suggested will have a n_client * model_size memory usage.
running sum will be 2 * model_size.

Reply:
I don't understand. that should not change the memory consumption at all but just the coding refactors